minidump-unwind/src/riscv64.rs

```diff
@@ -0,0 +1,313 @@
+
+            let mut valid = HashSet::new();
+            valid.insert(PROGRAM_COUNTER);
+            valid.insert(STACK_POINTER);
+
+            let context = MinidumpContext {
+                raw: MinidumpRawContext::Riscv64(caller_ctx),
+                valid: MinidumpContextValidity::Some(valid),
+            };
+            return Some(StackFrame::from_context(context, FrameTrust::Scan));
+        }
+    }
+
+    None
+}
+
+async fn instruction_seems_valid<P>(
+    instruction: Pointer,
+    modules: &MinidumpModuleList,
+    symbol_provider: &P,
+) -> bool
+where
+    P: SymbolProvider + Sync,
+{
+    super::instruction_seems_valid_by_symbols(instruction, modules, symbol_provider).await
+}
+
+#[async_trait::async_trait]
+impl Unwind for RiscvContext {
+    async fn get_caller_frame<P>(
+        &self,
+        callee: &StackFrame,
+        grand_callee: Option<&StackFrame>,
+        stack_memory: Option<UnifiedMemory<'_, '_>>,
+        modules: &MinidumpModuleList,
+        _system_info: &SystemInfo,
+        syms: &P,
+    ) -> Option<StackFrame>
+    where
+        P: SymbolProvider + Sync,
+    {
+        let stack = stack_memory?;
+
+        // .await doesn't like closures, so don't use Option chaining
+        let mut frame = None;
+        if frame.is_none() {
+            frame = get_caller_by_cfi(self, callee, grand_callee, stack, modules, syms).await;
+        }
+        if frame.is_none() {
+            frame = get_caller_by_frame_pointer(self, callee, grand_callee, stack, modules, syms);
+        }
+        if frame.is_none() {
+            frame = get_caller_by_scan(self, callee, stack, modules, syms).await;
+        }
+        let mut frame = frame?;
+
+        // We now check the frame to see if it looks like unwinding is complete,
+        // based on the frame we computed having a nonsense value. Returning
+        // None signals to the unwinder to stop unwinding.
+
+        // if the instruction is within the first ~page of memory, it's basically
+        // null, and we can assume unwinding is complete.
+        if frame.context.get_instruction_pointer() < 4096 {
+            trace!("instruction pointer was nullish, assuming unwind complete");
+            return None;
+        }
+
+        // If the new stack pointer is at a lower address than the old,
+        // then that's clearly incorrect. Treat this as end-of-stack to
+        // enforce progress and avoid infinite loops.
+
+        let sp = frame.context.get_stack_pointer();
+        let last_sp = self.get_register_always("sp");
+        if sp <= last_sp {
+            // Arm leaf functions may not actually touch the stack (thanks
+            // to the link register allowing you to "push" the return address
+            // to a register), so we need to permit the stack pointer to not
+            // change for the first frame of the unwind. After that we need
+            // more strict validation to avoid infinite loops.
```

Comment:
Likewise this looks like a copy-paste error, does the logic here hold true for RISC-V too? I'm not sufficiently familiar with the ABI to be able to tell.